src/olympia/devhub/forms.py

```diff
@@ -432,14 +433,25 @@ class Meta:
         model = ApplicationsVersions
         fields = ('application', 'min', 'max')
 
-    def __init__(self, *args, **kw):
-        super(CompatForm, self).__init__(*args, **kw)
+    def __init__(self, *args, **kwargs):
+        # 'version' should always be passed as a kwarg to this form. If it's
+        # absent, it probably means form_kwargs={'version': version} is missing
+        # from the instantiation of the formset.
+        version = kwargs.pop('version')
+        super(CompatForm, self).__init__(*args, **kwargs)
         if self.initial:
             app = self.initial['application']
         else:
             app = self.data[self.add_prefix('application')]
         self.app = amo.APPS_ALL[int(app)]
         qs = AppVersion.objects.filter(application=app).order_by('version_int')
+
+        # Legacy extensions can't set compatibility higher than 56.* for
+        # Firefox and Firefox for Android.
+        if (self.app in (amo.FIREFOX, amo.ANDROID) and
+                not version.is_webextension and
+                version.addon.type not in amo.NO_COMPAT):
+            qs = qs.filter(version_int__lt=57000000000000)
```

Comment:
current nightly `57.0a1` has a `version_int` of `57000000001100` so this should work perfectly.